sdk/storage/storage-blob/src/policies/StorageBearerTokenChallengeAuthenticationPolicy.ts

```diff
@@ -0,0 +1,423 @@
+  if (response.status === 401 && challenge) {
+    return challenge;
+  }
+  return;
+}
+
+/**
+ * Challenge structure
+ */
+interface Challenge {
+  authorization_uri: string;
+  resource_id: string;
+}
+
+/**
+ * Converts: `Bearer a="b" c="d"`.
+ * Into: `[ { a: 'b', c: 'd' }]`.
+ *
+ * @internal
+ */
+function parseChallenge(challenge: string): any {
+  const bearerChallenge = challenge.slice("Bearer ".length);
+  const challengeParts = `${bearerChallenge.trim()} `.split(" ").filter((x) => x);
+  const keyValuePairs = challengeParts.map((keyValue) =>
+    (([key, value]) => ({ [key]: value }))(keyValue.trim().split("="))
+  );
+  // Key-value pairs to plain object:
+  return keyValuePairs.reduce((a, b) => ({ ...a, ...b }), {});
+}
+
+// #endregion
+
+/**
+ * Creates a new factory for a RequestPolicy that applies a bearer token to
+ * the requests' `Authorization` headers.
+ *
+ * @param credential - The TokenCredential implementation that can supply the bearer token.
+ * @param scopes - The scopes for which the bearer token applies.
+ */
+
+export function storageBearerTokenChallengeAuthenticationPolicy(
+  credential: TokenCredential,
+  scopes: string | string[]
+): RequestPolicyFactory {
+  // This simple function encapsulates the entire process of reliably retrieving the token
+  let getToken = createTokenCycler(credential, scopes /* , options */);
+
+  class StorageBearerTokenChallengeAuthenticationPolicy extends BaseRequestPolicy {
+    public constructor(nextPolicy: RequestPolicy, options: RequestPolicyOptions) {
+      super(nextPolicy, options);
+    }
+
+    public async sendRequest(webResource: WebResourceLike): Promise<HttpOperationResponse> {
+      if (!webResource.url.toLowerCase().startsWith("https://")) {
+        throw new Error(
+          "Bearer token authentication is not permitted for non-TLS protected (non-https) URLs."
+        );
+      }
+
+      const getTokenInternal = getToken;
+      const token = (
+        await getTokenInternal({
+          abortSignal: webResource.abortSignal,
+          tracingOptions: {
+            tracingContext: webResource.tracingContext,
+          },
+        })
+      ).token;
+      webResource.headers.set(Constants.HeaderConstants.AUTHORIZATION, `Bearer ${token}`);
+
+      const response = await this._nextPolicy.sendRequest(webResource);
+
+      if (response?.status === 401) {
+        const challenge = getChallenge(response);
+        if (challenge) {
+          const challengeInfo: Challenge = parseChallenge(challenge);
+          const challengeScopes = challengeInfo.resource_id + Constants.DefaultScope;
+          const parsedAuthUri = URLBuilder.parse(challengeInfo.authorization_uri);
+          const pathSegments = parsedAuthUri.getPath()!.split("/");
+          const tenantId = pathSegments[1];
+          const getTokenForChallenge = createTokenCycler(
+            credential,
+            challengeScopes /* , options */
```

Comment:
Does the options still require?

Reply:
BearerTokenChallengeAuthenticationPolicy didn't specify option, we use the same code.